### ISSUE TITLE
accel_dpdk_cryptodev.c/accel_dpdk_cryptodev_process_task func will cause io hang when reach branch num_enqueued_ops < cryop_cnt and  num_enqueued_ops !=0

### ISSUE THREAD

**Issue by llicety** [spdk/spdk#2907]
# Sighting report

accel_dpdk_cryptodev.c/accel_dpdk_cryptodev_process_task func will cause io hang after rte_cryptodev_enqueue_burst called and next logic reach branch num_enqueued_ops < cryop_cnt and  num_enqueued_ops !=0.

when use openssl crypto driver, crypto ops processed in file rte_openssl_pmd.c/process_op func. if crypto ops processed success but rte_ring_enqueue failed in the end of process_op func, the accel_dpdk_cryptodev.c/accel_dpdk_cryptodev_process_task func will reach logic branch num_enqueued_ops < cryop_cnt and  num_enqueued_ops !=0 in accel_dpdk_cryptodev.c/accel_dpdk_cryptodev_process_task func, because the crypto_ops[num_enqueued_ops]->status ==  RTE_CRYPTO_OP_STATUS_SUCCESS

## Expected Behavior

if  next logic reach branch num_enqueued_ops < cryop_cnt and  num_enqueued_ops !=0 in accel_dpdk_cryptodev_process_task after rte_cryptodev_enqueue_burst called, it should set task->cryop_submitted = num_enqueued_ops and goto err_free_ops.

## Current Behavior

now we missed this logic branch. 
so in the accel_dpdk_cryptodev_poll_qp func, task->cryop_completed will never equal to task->cryop_submitted. this io will be hang here.

## Possible Solution

set task->cryop_submitted = num_enqueued_ops and then goto err_free_ops to free ops. the failed io will be complete by spdk_accel_task_complete and ops alse freed in accel_dpdk_cryptodev_process_task。

another choice may be retry the not enqueued_ops like case RTE_CRYPTO_OP_STATUS_NOT_PROCESSED.

## Steps to Reproduce

1. set ACCEL_DPDK_CRYPTODEV_QP_DESCRIPTORS less than task->cryop_submitted
2.
3.
4.

**Comment by AlekseyMarchuk** [comment]
Hi @llicety 
I would like to clarify - is the following really true `crypto_ops[num_enqueued_ops]->status == RTE_CRYPTO_OP_STATUS_SUCCESS`? It looks weird, but if you do see this behaviour then I'll take it into account.
Will push a fix tomorrow

**Comment by AlekseyMarchuk** [comment]
Please try this patch https://review.spdk.io/gerrit/c/spdk/spdk/+/16784
I think that setting `task->cryop_submitted` to the number of submitted crypto operations is enough

**Comment by llicety** [comment]
> Hi @llicety I would like to clarify - is the following really true `crypto_ops[num_enqueued_ops]->status == RTE_CRYPTO_OP_STATUS_SUCCESS`? It looks weird, like a bug in openssl pmd, but if you do see this behaviour then I'll take it into account. Also, we can't use `goto err_free_ops` if we submitted at least 1 crypto op - mbufs and cryop structs might be used by HW/SW driver and we'll later reap a completion for this ops and free resources again.

Hi @AlekseyMarchuk  Yes, crypto_ops[num_enqueued_ops]->status == RTE_CRYPTO_OP_STATUS_SUCCESS. I confirmed this in my project based on spdk v21.04 and I checked the latest spdk version, the logic are the same. the gdb info as follow:`Breakpoint 3, _crypto_operation (bdev_io=0x7f5e32ee8100, crypto_op=RTE_CRYPTO_CIPHER_OP_DECRYPT, aux_buf=0x0) at vbdev_crypto.c:1101
1101    in vbdev_crypto.c
(gdb) p num_enqueued_ops
$3 = 63
(gdb) p cryop_cnt
$4 = 168
(gdb) p crypto_ops[num_enqueued_ops]->status
$5 = 0 '\000'
`

**Comment by llicety** [comment]
> Please try this patch https://review.spdk.io/gerrit/c/spdk/spdk/+/16784 I think that setting `task->cryop_submitted` to the number of submitted crypto operations is enough

only set 

> Please try this patch https://review.spdk.io/gerrit/c/spdk/spdk/+/16784 I think that setting `task->cryop_submitted` to the number of submitted crypto operations is enough

only set task->cryop_submitted to the number of submitted crypto operations is not enough.  In this way accel_dpdk_cryptodev_poll_qp func only free num_dequeued_ops ops and the related mbuf. The ops and buf allocated but enqueue faild will not free.

**Comment by llicety** [comment]
> > Please try this patch https://review.spdk.io/gerrit/c/spdk/spdk/+/16784 I think that setting `task->cryop_submitted` to the number of submitted crypto operations is enough
> 
> only set
> 
> > Please try this patch https://review.spdk.io/gerrit/c/spdk/spdk/+/16784 I think that setting `task->cryop_submitted` to the number of submitted crypto operations is enough
> 
> only set task->cryop_submitted to the number of submitted crypto operations is not enough. In this way accel_dpdk_cryptodev_poll_qp func only free num_dequeued_ops ops and the related mbuf. The ops and buf allocated but enqueue faild will not free.

**Comment by AlekseyMarchuk** [comment]
@llicety Thank you, I'll update the patch soon. I think that handling of cryops with status `RTE_CRYPTO_OP_STATUS_NOT_PROCESSED` needs to be reworked - we can just return mbufs and cryops to the pool and completely remove queuing of such ops. Ops which were not submitted will be handled later when we reap completions

**Comment by AlekseyMarchuk** [comment]
@llicety I updated this patch https://review.spdk.io/gerrit/c/spdk/spdk/+/16784
Could you verify it?

**Comment by llicety** [comment]
> @llicety I updated this patch https://review.spdk.io/gerrit/c/spdk/spdk/+/16784 Could you verify it?

@AlekseyMarchuk thank you very much for this patch. This patch can solved the problem, but it may cause another problem when task->inplace is true (decrypt io). 
When decrypt io, task->inplace is true, crypto_ops[num_enqueued_ops] ops src buf data has been decrypted but enqueue ring failed, then it will be put to queued_tasks. When process  queued_tasks, the crypto_ops[num_enqueued_ops] ops src buf data decrypted will be decrypted again .  Is it possible to cause a data consistency problem?

**Comment by AlekseyMarchuk** [comment]
Hi @llicety sorry for long delay
Will it work for you if we treat `RTE_CRYPTO_OP_STATUS_SUCCESS ` as error case? It can be a quick fix of original problem where IO hangs. I can later fix handling of success status

**Comment by AlekseyMarchuk** [comment]
@llicety 
I updated this patch https://review.spdk.io/gerrit/c/spdk/spdk/+/16784 to remove queued_cry_ops list. Crypto op with status `RTE_CRYPTO_OP_STATUS_SUCCESS` is handled as an error. The next patch https://review.spdk.io/gerrit/c/spdk/spdk/+/16995 handles `RTE_CRYPTO_OP_STATUS_SUCCESS` and it needs your verification. Could you please check it?

**Comment by AlekseyMarchuk** [comment]
@llicety could you verify these two patches?
https://review.spdk.io/gerrit/c/spdk/spdk/+/16784/
https://review.spdk.io/gerrit/c/spdk/spdk/+/16995

**Comment by llicety** [comment]
> @llicety could you verify these two patches? https://review.spdk.io/gerrit/c/spdk/spdk/+/16784/ https://review.spdk.io/gerrit/c/spdk/spdk/+/16995

Hi @AlekseyMarchuk，thanks for all the effort you had done for this issus.
I can't verify these two patch in may project now.  I reviewed these patch, it's good ideal. but return in accel_dpdk_cryptodev_poller func when rc == -EINPROGRESS will lose queued_tasks_tmp, it may be a problem.

**Comment by AlekseyMarchuk** [comment]
@llicety I don't see any return in accel_dpdk_cryptodev_poller when handling -EINPROGRESS, queued_tasks_tmp list should be processed correctly. 
There was another issue #2759 related to this problem submitted by @yin19941005. @yin19941005 please loot at 2 patches which should fix this issue:
https://review.spdk.io/gerrit/c/spdk/spdk/+/16784/ 
https://review.spdk.io/gerrit/c/spdk/spdk/+/16995

**Comment by yin19941005** [comment]
Hi @AlekseyMarchuk,

Unfortunately, the patches doesn't fix the issue. I am going to share the details with you:

### Setup
 
**Git logs** of the patch applied:
```
commit db6d1d50fee3707c599978690e17e18ea6e70bd5 (HEAD -> change-16995)
Author: Alexey Marchuk <[alexeymar@nvidia.com](mailto:alexeymar@nvidia.com)>
Date:   Wed Mar 1 10:04:01 2023 +0100

    [TEST] accel/dpdk_cryptodev: Handle OP_STATUS_SUCCESS

    SW PMD might process a crypto operation but failed
    to submit it to a completions ring.
    Such operation can't be retried if crypto operation
    is inplace.
    Handle such crypto op as a completed.

    Signed-off-by: Alexey Marchuk <[alexeymar@nvidia.com](mailto:alexeymar@nvidia.com)>
    Change-Id: Ida161cec045167af752ebd5b57f41b2bbfe8b97c

commit 67670ec825b33ae8cf6536ca0fcc27520746677b (change-16784)
Author: Alexey Marchuk <[alexeymar@nvidia.com](mailto:alexeymar@nvidia.com)>
Date:   Sun Feb 12 16:06:26 2023 +0100

    accel/dpdk_cryptodev: Remove queued_cry_ops

    If we were not able to submit all configured
    crypto ops, then we can just release crypto_ops
    and mbuf object of these crypto ops and save
    the actual number of submitted operation in
    the accel task. Once all submitted operations
    complete, poller will call
    accel_dpdk_cryptodev_process_task func to submit
    cyrpto operations for reamining data blocks.
    If no crypto ops were submitted then the task
    will be palced in the channel's queued_tasks
    and poller will try to resubmit the task.
    That in theory should increase performance
    since we attempted to resubmit queued ops
    with burst size==1 which is not efficient

    Fixes issue #2907

    Signed-off-by: Alexey Marchuk <[alexeymar@nvidia.com](mailto:alexeymar@nvidia.com)>
    Change-Id: I4d17e8ed1ad5383848e4d09c46009c6cb2834360

commit 4037f0671b4759449a9f3f50dec3d397ef908f96
Author: Jim Harris <[james.r.harris@intel.com](mailto:james.r.harris@intel.com)>
Date:   Thu Feb 23 14:11:36 2023 +0000

    sock/posix: don't add 1 to pipe size

    There's no need to add 1 to the pipe size, it
    actually results in unaligned accesses after we've
    cycled through the pipe buffer for the first time.
```

**Compile**
```
$ ./configure --disable-tests --disable-unit-tests --with-crypto --without-fio --with-vhost \
--without-pmdk --without-rbd --with-rdma --with-shared --with-iscsi-initiator --enable-debug
$ make
```

The SPDK **NVMe target** setup:
```
Run nvmf target:
nvmf_tgt -m 0xf0 -L vbdev_crypto --wait-for-rpc --env-context "--allow=0000:01:00.0,class=crypto,wcs_file=/etc/mellanox/credential.txt"

Configure:
rpc.py dpdk_cryptodev_scan_accel_module
rpc.py dpdk_cryptodev_set_driver -d mlx5_pci
rpc.py accel_assign_opc -o encrypt -m dpdk_cryptodev
rpc.py accel_assign_opc -o decrypt -m dpdk_cryptodev
rpc.py framework_start_init
rpc.py bdev_malloc_create -b Malloc1 4096 512
rpc.py accel_crypto_key_create -c AES_XTS -e eb38f46028dfb7d58675054d13a97430 -k b2b1aefc4d7f42b0530aa48841b026030a8e00ca1e932bd4a13cb9ab48954426 -n sample_key
rpc.py bdev_crypto_create -n sample_key Malloc1 SS_Malloc1

```

The **initiator** connect thru 100Gbe RDMA (RHEL 8 native NVMe host)


### Test

#### Write

**Initiator** run:
```
$ dd if=/dev/zero of=/dev/nvme2n1 bs=1M count=10
10+0 records in
10+0 records out
10485760 bytes (10 MB, 10 MiB) copied, 0.0174932 s, 599 MB/s
```

Then ERROR message on **target**:
```
[2023-03-28 16:38:00.803304] vbdev_crypto.c: 125:_crypto_operation_complete: *ERROR*: Issue with encryption on bdev_io 0x200015d438c0
[2023-03-28 16:38:00.803386] vbdev_crypto.c: 125:_crypto_operation_complete: *ERROR*: Issue with encryption on bdev_io 0x200015d45300
[2023-03-28 16:38:00.804395] vbdev_crypto.c: 125:_crypto_operation_complete: *ERROR*: Issue with encryption on bdev_io 0x200015d47d00
[2023-03-28 16:38:00.804410] vbdev_crypto.c: 125:_crypto_operation_complete: *ERROR*: Issue with encryption on bdev_io 0x200015d47280
[2023-03-28 16:38:00.808153] vbdev_crypto.c: 125:_crypto_operation_complete: *ERROR*: Issue with encryption on bdev_io 0x200015cf5c80
```

Hexdump on **Initiator** to verify:
```
Host side see random data for random address (after zero the whole disk):
hexdump -C /dev/nvme2n1 -n 330000
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0004b400  99 54 ef 89 4a 85 ea dc  13 00 44 0d 0c e9 2a 39  |.T..J.....D...*9|
0004b410  74 59 59 88 ea 45 71 83  f7 c0 f1 61 13 af c7 b9  |tYY..Eq....a....|
0004b420  5a 0d 0d b3 95 0f 54 7a  22 ff 95 5a 7f 95 fa b3  |Z.....Tz"..Z....|
```

#### Read
seems Read normal but target has error message:

**Initiator** run:
```
$ dd if=/dev/nvme2n1 of=/dev/null bs=1M count=2
2+0 records in
2+0 records out
2097152 bytes (2.1 MB, 2.0 MiB) copied, 0.0103968 s, 202 MB/s
```

ERROR message on target:
```
[2023-03-28 16:41:23.757519] vbdev_crypto.c: 109:_crypto_operation_complete: *ERROR*: Issue with decryption on bdev_io 0x200015ceee40
[2023-03-28 16:41:23.760574] vbdev_crypto.c: 109:_crypto_operation_complete: *ERROR*: Issue with decryption on bdev_io 0x200015caba00
[2023-03-28 16:41:23.761053] vbdev_crypto.c: 109:_crypto_operation_complete: *ERROR*: Issue with decryption on bdev_io 0x200015ca9000
[2023-03-28 16:41:23.761523] vbdev_crypto.c: 109:_crypto_operation_complete: *ERROR*: Issue with decryption on bdev_io 0x200015ccacc0
```

When I try to read more data for verify, core dump on target.
**Initiator** read more data:
```
$ dd if=/dev/nvme2n1 of=/dev/null bs=1M count=8
```

ERROR on **target**:
```
[2023-03-29 11:26:16.371290] vbdev_crypto.c: 109:_crypto_operation_complete: *ERROR*: Issue with decryption on bdev_io 0x200015cede80
[2023-03-29 11:26:16.371335] vbdev_crypto.c: 109:_crypto_operation_complete: *ERROR*: Issue with decryption on bdev_io 0x200015cda380
[2023-03-29 11:26:16.371608] vbdev_crypto.c: 109:_crypto_operation_complete: *ERROR*: Issue with decryption on bdev_io 0x200015cac480
[2023-03-29 11:26:16.371623] vbdev_crypto.c: 109:_crypto_operation_complete: *ERROR*: Issue with decryption on bdev_io 0x200015caba00
[2023-03-29 11:26:16.371635] vbdev_crypto.c: 109:_crypto_operation_complete: *ERROR*: Issue with decryption on bdev_io 0x200015caaf80
[2023-03-29 11:26:16.371646] vbdev_crypto.c: 109:_crypto_operation_complete: *ERROR*: Issue with decryption on bdev_io 0x200015ca9a80
./start-nvmf-tgt.sh: line 24:  2064 Segmentation fault      (core dumped) $SPDK_PATH/bin/nvmf_tgt $SPDK_OPT --env-context "$ENV_CTX"


Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007fe82d1a16a1 in nvmf_request_get_buffers (req=0x2001203d0d68, group=0x7fe818198120, transport=0x12baa60,
    length=24576, io_unit_size=8192, cb_func=0x7fe82d1a1524 <nvmf_request_set_buffer>) at transport.c:816
816                             STAILQ_REMOVE_HEAD(&group->buf_cache, link);
[Current thread is 1 (Thread 0x7fe8216f2700 (LWP 2066))]
```

The Core dump file:
[core.reactor_4.0.f5473423834141ffbc01937d45291fc1.2064.1680114376000000.gz](https://github.com/spdk/spdk/files/11105658/core.reactor_4.0.f5473423834141ffbc01937d45291fc1.2064.1680114376000000.gz)

@AlekseyMarchuk, could you take a look on the core dump file to check what's the error? I had tested the crypto_aesni_mb as well, same error and the same line in the coredump.

**Comment by AlekseyMarchuk** [comment]
@yin19941005 thank you for response and detailed description. I'll try aesni_mb on my setup. I tested with 128KiB of data, maybe using 1MiB leads to a different result

**Comment by AlekseyMarchuk** [comment]
@yin19941005 I fixed this issue in https://review.spdk.io/gerrit/c/spdk/spdk/+/17456
Please try this series https://review.spdk.io/gerrit/c/spdk/spdk/+/16995
I also verified the last patch on my setup by adding openssl driver support

**Comment by yin19941005** [comment]
Hi @AlekseyMarchuk, 

Thank you for helping! The latest patches works on my environment! I tested with three crypto driver: aesni_mb, mlx5 and openssl. All of them are working fine: no segmentation fault, no random data and no error message read/write. I tried to run the dd command write a file to the drive and was trying to compare SHA2 checksum, then I found the checksum is matched as well. I think is safe to say the issue is fixed
